int-set/src/bitpage.rs

```diff
@@ -0,0 +1,353 @@
+        let last = last & PAGE_MASK;
+        let first_elem_idx = first / ELEM_BITS;
+        let last_elem_idx = last / ELEM_BITS;
+
+        for elem_idx in first_elem_idx..=last_elem_idx {
+            let elem_start = first.max(elem_idx * ELEM_BITS) & ELEM_MASK;
+            let elem_last = last.min(((elem_idx + 1) * ELEM_BITS) - 1) & ELEM_MASK;
+
+            let end_shift = ELEM_BITS - elem_last - 1;
+            let mask = u64::MAX << (elem_start + end_shift);
+            let mask = mask >> end_shift;
+
+            self.storage[elem_idx as usize] |= mask;
+        }
+
+        self.mark_dirty();
+    }
+
+    /// Removes (val % page width) from this set.
+    pub fn remove(&mut self, val: u32) -> bool {
+        let ret = self.contains(val);
+        *self.element_mut(val) &= !elem_index_bit_mask(val);
+        self.mark_dirty();
+        ret
+    }
+
+    /// Return true if (val % page width) is a member of this set.
+    pub fn contains(&self, val: u32) -> bool {
+        (*self.element(val) & elem_index_bit_mask(val)) != 0
+    }
+
+    fn mark_dirty(&mut self) {
+        self.len.set(u32::MAX);
+    }
+
+    fn is_dirty(&self) -> bool {
+        self.len.get() == u32::MAX
+    }
+
+    fn element(&self, value: u32) -> &Element {
+        let idx = (value & PAGE_MASK) / ELEM_BITS;
+        &self.storage[idx as usize]
+    }
+
+    fn element_mut(&mut self, value: u32) -> &mut Element {
+        let idx = (value & PAGE_MASK) / ELEM_BITS;
+        &mut self.storage[idx as usize]
+    }
+}
+
+/// returns the bit to set in an element for this value
+const fn elem_index_bit_mask(value: u32) -> Element {
+    1 << (value & ELEM_MASK)
+}
+
+fn iter_bit_indices(val: Element) -> impl Iterator<Item = u32> {
+    let mut idx = 0;
+
+    std::iter::from_fn(move || {
+        if idx >= ELEM_BITS {
+            return None;
+        }
+        let mask = 1u64.checked_shl(idx).unwrap_or(0) - 1;
+        let masked = val & !mask;
+        let next_index = masked.trailing_zeros();
+        if next_index >= ELEM_BITS {
+            return None;
+        }
+        idx = next_index + 1;
+        Some(next_index)
+    })
+}
+
+impl Default for BitPage {
+    fn default() -> Self {
+        Self::new_zeroes()
+    }
+}
+
+impl std::fmt::Debug for BitPage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        let values: Vec<_> = self.iter().collect();
+        std::fmt::Debug::fmt(&values, f)
+    }
+}
+
+impl Hash for BitPage {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.storage.hash(state);
+    }
+}
+
+impl std::cmp::PartialEq for BitPage {
+    fn eq(&self, other: &Self) -> bool {
+        self.storage == other.storage
+    }
+}
+
+impl std::cmp::Eq for BitPage {}
+
+#[cfg(test)]
+mod test {
+    use std::collections::HashSet;
+
+    use super::*;
+
+    impl BitPage {
+        /// Create a new page with all bits set.
+        pub fn new_ones() -> Self {
```

Comment:
is this down here because you were getting a dead code warning?

Reply:
Yes, we currently only use it in tests. We may eventually need it for code in the public api, at that time we can move it back up.